### PR TITLE
[Workspace] Add User level setting

### PR DESCRIPTION
**Pull request by Hailong-am**
### Description

This PR includes 
1. user settings page
2. ui_setting_client changes to support user personal settings
3. UserUISettingsClientWrapper to translate user id placeholder to real user id

### Issues Resolved

https://github.com/opensearch-project/OpenSearch-Dashboards/issues/7821

https://github.com/opensearch-project/OpenSearch-Dashboards/issues/7909

## Screenshot

https://github.com/user-attachments/assets/57e9962b-4869-4a41-911e-9a698480708e

## Testing the changes

1. Run OSD enabled `yarn start --no-base-path --workspace.enabled=true`
2. make sure security is enabled
3. open http://localhost:5601, and then go to dev tools
4. Run kibana search to make sure global setting been created
```
GET .kibana/_search
{
  "query": {
    "bool": {
      "must": [
        {
          "match": {
            "type": "config"
          }
        }
      ]
    }
  }
```
5. Run command to add user setting, after that run kibana query again to verify 
a config doc with id `config:admin` been indexed and it has defaultWorkspace as "ABC"
```
curl --request POST \
  --url http://localhost:5601/api/opensearch-dashboards/settings \
  --header 'Authorization: Basic YWRtaW46YWRtaW4=' \
  --header 'Content-Type: application/json' \
  --header 'osd-xsrf: osd-fetch' \
  --data '{
 "changes": {
  "defaultIndex": "logstash-*",
  "defaultWorkspace": "ABC"
 }
}'
```

## Changelog

- feat: [Workspace] Support user personal settings

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/core/server/ui_settings/ui_settings_client.ts
Comment (SuZhou-Joe) [inline src/core/server/ui_settings/ui_settings_client.ts:68]:
```suggestion
export const CURRENT_USER_PLACEHOLDER = '<current_user>';
```
`CURRENT_USER` seems a little bit confusing to me, actually it is a placeholder if I understand correctly.

File: src/plugins/workspace/public/components/workspace_list/index.tsx
Comment (raintygao) [inline src/plugins/workspace/public/components/workspace_list/index.tsx:117]: Nit: Would it be more clear if it is named as `defaultWorkspaceId`?

File: src/plugins/workspace/public/components/workspace_list/index.tsx
Comment (raintygao) [inline src/plugins/workspace/public/components/workspace_list/index.tsx:477]: Do we need this config if it would always be unavailable?

File: src/core/server/ui_settings/ui_settings_client.ts
Comment (ruanyl) [inline src/core/server/ui_settings/ui_settings_client.ts:260]: I think a boolean flag `userLevel: boolean` may not flexible enough, imagine in the future, we may have uiSettings client to support workspace level settings.

File: src/plugins/workspace/public/components/workspace_list/index.tsx
Comment (Hailong-am) [reply src/plugins/workspace/public/components/workspace_list/index.tsx:477]: we plan to enable it in next PR to let user leave from selected workspace only if user permission granted by individual user id not group.

File: src/plugins/workspace/public/components/workspace_list/index.tsx
Comment (Hailong-am) [reply src/plugins/workspace/public/components/workspace_list/index.tsx:117]: thanks, will update to `defaultWorkspaceId`

File: src/core/server/ui_settings/ui_settings_client.ts
Comment (ruanyl) [reply src/core/server/ui_settings/ui_settings_client.ts:247]: Here is just a few changes to demonstrate what I was referring:
```diff
diff --git a/src/core/server/ui_settings/ui_settings_client.ts b/src/core/server/ui_settings/ui_settings_client.ts
index e78f797465..b48a32e6d9 100644
--- a/src/core/server/ui_settings/ui_settings_client.ts
+++ b/src/core/server/ui_settings/ui_settings_client.ts
@@ -50,6 +50,7 @@ export interface UiSettingsServiceOptions {
 interface ReadOptions {
   ignore401Errors?: boolean;
   autoCreateOrUpgradeIfMissing?: boolean;
+  scope?: Scope;
 }
 
 interface UserProvidedValue<T = unknown> {
@@ -62,6 +63,8 @@ type UiSettingsRawValue = UiSettingsParams & UserProvidedValue;
 type UserProvided<T = unknown> = Record<string, UserProvidedValue<T>>;
 type UiSettingsRaw = Record<string, UiSettingsRawValue>;
 
+type Scope = 'user' | 'global';
+
 export class UiSettingsClient implements IUiSettingsClient {
   private readonly type: UiSettingsServiceOptions['type'];
   private readonly id: UiSettingsServiceOptions['id'];
@@ -100,13 +103,13 @@ export class UiSettingsClient implements IUiSettingsClient {
     return this.defaults[key]?.value;
   }
 
-  async get<T = any>(key: string): Promise<T> {
-    const all = await this.getAll();
+  async get<T = any>(key: string, scope?: Scope): Promise<T> {
+    const all = await this.getAll(scope);
     return all[key];
   }
 
-  async getAll<T = any>() {
-    const raw = await this.getRaw();
+  async getAll<T = any>(scope?: Scope) {
+    const raw = await this.getRaw(scope);
 
     return Object.keys(raw).reduce((all, key) => {
       const item = raw[key];
@@ -115,8 +118,10 @@ export class UiSettingsClient implements IUiSettingsClient {
     }, {} as Record<string, T>);
   }
 
-  async getUserProvided<T = unknown>(): Promise<UserProvided<T>> {
-    const userProvided: UserProvided<T> = this.onReadHook<T>(await this.read());
+  async getUserProvided<T = unknown>(scope?: Scope): Promise<UserProvided<T>> {
+    const userProvided: UserProvided<T> = this.onReadHook<T>(
+      await this.read({ scope, autoCreateOrUpgradeIfMissing: scope === 'global' })
+    );
 
     // write all overridden keys, dropping the userValue is override is null and
     // adding keys for overrides that are not in saved object
@@ -128,13 +133,13 @@ export class UiSettingsClient implements IUiSettingsClient {
     return userProvided;
   }
 
-  async setMany(changes: Record<string, any>) {
+  async setMany(changes: Record<string, any>, scope?: Scope) {
     this.onWriteHook(changes);
-    await this.write({ changes });
+    await this.write({ changes, scope });
   }
 
-  async set(key: string, value: any) {
-    await this.setMany({ [key]: value });
+  async set(key: string, value: any, scope?: Scope) {
+    await this.setMany({ [key]: value }, scope);
   }
 
   async remove(key: string) {
@@ -159,8 +164,8 @@ export class UiSettingsClient implements IUiSettingsClient {
     }
   }
 
-  private async getRaw(): Promise<UiSettingsRaw> {
-    const userProvided = await this.getUserProvided();
+  private async getRaw(scope?: Scope): Promise<UiSettingsRaw> {
+    const userProvided = await this.getUserProvided(scope);
     return defaultsDeep(userProvided, this.defaults);
   }
 
@@ -204,25 +209,32 @@ export class UiSettingsClient implements IUiSettingsClient {
   private async write({
     changes,
     autoCreateOrUpgradeIfMissing = true,
+    scope = 'global',
   }: {
     changes: Record<string, any>;
     autoCreateOrUpgradeIfMissing?: boolean;
+    scope?: Scope;
   }) {
     changes = this.translateChanges(changes, 'timeline', 'timelion');
+    const id = this.getId(scope);
     try {
-      await this.savedObjectsClient.update(this.type, this.id, changes);
+      await this.savedObjectsClient.update(this.type, id, changes);
     } catch (error) {
       if (!SavedObjectsErrorHelpers.isNotFoundError(error) || !autoCreateOrUpgradeIfMissing) {
         throw error;
       }
 
-      await createOrUpgradeSavedConfig({
-        savedObjectsClient: this.savedObjectsClient,
-        version: this.id,
-        buildNum: this.buildNum,
-        log: this.log,
-        handleWriteErrors: false,
-      });
+      if (scope === 'global') {
+        await createOrUpgradeSavedConfig({
+          savedObjectsClient: this.savedObjectsClient,
+          version: this.id,
+          buildNum: this.buildNum,
+          log: this.log,
+          handleWriteErrors: false,
+        });
+      } else {
+        await this.savedObjectsClient.create('config', {}, { id });
+      }
 
       await this.write({
         changes,
@@ -231,12 +243,21 @@ export class UiSettingsClient implements IUiSettingsClient {
     }
   }
 
+  getId(scope?: Scope) {
+    if (scope === 'user') {
+      return '__USER_SETTING_FLAG__';
+    }
+    return this.id;
+  }
+
   private async read({
     ignore401Errors = false,
     autoCreateOrUpgradeIfMissing = true,
+    scope = 'global',
   }: ReadOptions = {}): Promise<Record<string, any>> {
+    const id = this.getId(scope);
     try {
-      const resp = await this.savedObjectsClient.get<Record<string, any>>(this.type, this.id);
+      const resp = await this.savedObjectsClient.get<Record<string, any>>(this.type, id);
       return this.translateChanges(resp.attributes, 'timelion', 'timeline');
     } catch (error) {
       if (SavedObjectsErrorHelpers.isNotFoundError(error) && autoCreateOrUpgradeIfMissing) {
@@ -262,6 +283,10 @@ export class UiSettingsClient implements IUiSettingsClient {
         return {};
       }
 
+      if (scope === 'user') {
+        return {};
+      }
+
       throw error;
     }
   }
```

File: src/plugins/advanced_settings/public/management_app/advanced_settings.tsx
Comment (SuZhou-Joe) [inline src/plugins/advanced_settings/public/management_app/advanced_settings.tsx:175]: Can we add an enum for scope?

File: src/core/server/ui_settings/ui_settings_client.ts
Comment (Hailong-am) [reply src/core/server/ui_settings/ui_settings_client.ts:68]: make sense, will updated to suggested name.

File: src/plugins/advanced_settings/public/plugin.ts
Comment (SuZhou-Joe) [inline src/plugins/advanced_settings/public/plugin.ts:194]: userSettingsEnabled?

File: src/plugins/advanced_settings/public/plugin.ts
Comment (SuZhou-Joe) [inline src/plugins/advanced_settings/public/plugin.ts:132]:
```suggestion
```
No need to registerAppUpdater for all of the apps because you have already add the updater$ when register the application.

File: src/plugins/advanced_settings/public/plugin.ts
Comment (SuZhou-Joe) [inline src/plugins/advanced_settings/public/plugin.ts:162]:
```suggestion
          order: 101, // user_settings should sit between `application settings` and `Assets`.
```

File: src/plugins/advanced_settings/public/plugin.ts
Comment (SuZhou-Joe) [inline src/plugins/advanced_settings/public/plugin.ts:139]: Nit: the id is used in 3 places, better to assign to a variable.

File: src/plugins/advanced_settings/server/plugin.ts
Comment (SuZhou-Joe) [inline src/plugins/advanced_settings/server/plugin.ts:78]:
```suggestion
      3, // The wrapper should be triggered after workspace_id_consumer wrapper to avoid creating user settings within any workspace.
```

File: src/plugins/advanced_settings/server/plugin.ts
Comment (SuZhou-Joe) [inline src/plugins/advanced_settings/server/plugin.ts:84]: There is a getPrincipalFromRequest util in #7959 and we can try using that util to get user_name.

File: src/core/server/ui_settings/ui_settings_client.ts
Comment (Hailong-am) [reply src/core/server/ui_settings/ui_settings_client.ts:247]: > What if a setting is both `global` and `user`? This implementation implies a setting can only be `global` or `user`.
> 
> This is what I proposed in the RFC, any reason why this didn't work?
> 
> ## Reading a setting with scope
> When calling uiSettings.get(key), it returns the value based on the following rules:
> 
> 1. If the setting was registered with scope global or empty, the global setting will be returned.
> 2. If the setting was registered with scope user, the user setting will be returned
> 3. If the setting was registered with both user and global, the user setting will be returned if it's set, otherwise, return the global setting
> 4. If calling uiSettings.get(key, defaultValue, 'user') with scope specified, it will return the setting of the specified scope. If the value doesn't not exist, then return the defaultValue. This requires to update the current uiSettings service's interface
> 
> ## Writing a setting with scope
> When calling uiSettings.set(key, value), it follows the following rules:
> 
> 1. If the setting was registered with scope global or empty, it will write the value to global setting saved object.
> 2. If the setting was registered with scope user, it will write the value to user setting saved object.
> 3. If the setting was registered with both user and global, it will by default update the user setting saved object.
> 4. If calling uiSettings.set(key, value, 'global') with scope specified, it will write the value to the corresponding saved object. This requires to update the current uiSettings service's interface

If a setting is both `global` and `user`, it will set to user setting by default, when get it they will merge, and returns same value when set. 

This PR implementation didn't implement item 4 for both read and write with specified scope that requires client to know which setting is global and which is user. I agree that will give more flexibility to control how we are going to store the setting but that also need additional validation make sure changes are align with specified scope.


To support item 4, that will be similar change as you suggested plus router accept scope parameter and scope validation.

As the default behavior could handle most of case, that's the reason i didn't implement it this PR.

Comment (Hailong-am) [reply src/core/server/ui_settings/ui_settings_client.ts:297]: scope added to both get and set method

File: src/core/server/ui_settings/ui_settings_client.ts
Comment (Hailong-am) [reply src/core/server/ui_settings/ui_settings_client.ts]: update to scope as suggested, it's more flexible. thanks.

File: src/plugins/advanced_settings/public/plugin.ts
Comment (Hailong-am) [reply src/plugins/advanced_settings/public/plugin.ts]: variable name updated

File: src/core/server/ui_settings/create_or_upgrade_saved_config/create_or_upgrade_saved_config.ts
Comment (ruanyl) [inline src/core/server/ui_settings/create_or_upgrade_saved_config/create_or_upgrade_saved_config.ts:38]: Seems `create_or_upgrade_saved_config` and `ui_settings_client` import from each other, maybe move `generateDocId` to somewhere else?

File: src/plugins/advanced_settings/server/plugin.ts
Comment (Hailong-am) [reply src/plugins/advanced_settings/server/plugin.ts:84]: sure, i will copy the file to my PR at this moment to unblock

File: src/core/server/ui_settings/create_or_upgrade_saved_config/create_or_upgrade_saved_config.ts
Comment (Hailong-am) [reply src/core/server/ui_settings/create_or_upgrade_saved_config/create_or_upgrade_saved_config.ts]: good catch, move it to utils

File: src/plugins/advanced_settings/server/plugin.ts
Comment (Hailong-am) [reply src/plugins/advanced_settings/server/plugin.ts]: comments added

File: src/plugins/advanced_settings/public/management_app/advanced_settings.tsx
Comment (Hailong-am) [reply src/plugins/advanced_settings/public/management_app/advanced_settings.tsx]: sure, enum added

**Comment by Hailong-am** [comment]
two cypress test failure are related to data-explorer, not with this change

**Review by SuZhou-Joe — DISMISSED** [review]

File: src/plugins/advanced_settings/server/saved_objects/user_ui_settings_client_wrapper.ts
Comment (ruanyl) [inline src/plugins/advanced_settings/server/saved_objects/user_ui_settings_client_wrapper.ts:114]: Do we still need to add this `references`

Comment (Hailong-am) [reply src/plugins/advanced_settings/server/saved_objects/user_ui_settings_client_wrapper.ts]: It's used to differentiate it's user setting not application settings. It's not mandatory, remove it now.

File: src/core/server/ui_settings/utils.ts
Comment (ruanyl) [inline src/core/server/ui_settings/utils.ts:8]: I have an impression that the current `request` can injected into uiSettingClient when I was doing the investigation,  I will take another look, but I think that won't block this PR.

File: src/plugins/workspace/common/types.ts
Comment (ruanyl) [inline src/plugins/workspace/common/types.ts:31]: Perhaps we can use the `PermissionModeId` from `src/core/types/workspace.ts`?

File: src/core/types/workspace.ts
Comment (ruanyl) [inline src/core/types/workspace.ts:28]: Q: Is this the permission of current user to the workspace?

File: src/plugins/advanced_settings/public/management_app/advanced_settings.tsx
Comment (ruanyl) [inline src/plugins/advanced_settings/public/management_app/advanced_settings.tsx:176]:
```suggestion
      .filter(([, setting]) => setting.scope !== UiSettingScope.USER)
```

**Review by ruanyl — APPROVED** [review]

File: src/core/types/workspace.ts
Comment (Hailong-am) [reply src/core/types/workspace.ts:28]: Yes, it is permission for current user which shows here 
image

File: src/plugins/workspace/common/types.ts
Comment (Hailong-am) [reply src/plugins/workspace/common/types.ts:31]: I have trouble to combine them previously, don't remember the reason, but now it works. anyway, combine them into one.

**Review by SuZhou-Joe — APPROVED** [review]

**Review by ruanyl — APPROVED** [review]

**Comment by Hailong-am** [comment]
i will do manual backport due to 

```ts
 const enableUserControl = core.uiSettings.get('theme:enableUserControl');
    if (enableUserControl) {
      setupTopNavThemeButton(core, core.uiSettings.get('home:useNewHomePage'));
    }
```
 not available in 2.x

**Comment by AMoo-Miki** [comment]
Manually backported to 2.x with https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8057

**Pull request merged**
